controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+	from, err := semver.Parse(fromVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+	}
+	to, err := semver.Parse(toVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+	}
+	if from.Compare(to) >= 0 {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	}
+
+	// check if the from version is even in the list of coredns versions
+	if _, ok := migration.Versions[fmt.Sprintf("%d.%d.%d", from.Major, from.Minor, from.Patch)]; !ok {
+		return fmt.Errorf("fromVersion %q is not a compatible coredns version", from.String())
+	}
+	return nil
+}
+
+// getCoreDNSInfo gets
+// - the current CoreDNS image version installed
+// - the current Corefile Configuration of CoreDNS
+// - the coredns deployment
+func (w *Workload) getCoreDNSInfo(ctx context.Context) (*coreDNSInfo, error) {
+	// Get the coredns configmap and corefile.
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	cm := &corev1.ConfigMap{}
+	if err := w.Client.Get(ctx, corednsKey, cm); err != nil {
+		return nil, errors.Wrapf(err, "error getting %s config map from target cluster", corednsKey.String())
+	}
+	corefile, ok := cm.Data[corefileKey]
+	if !ok {
+		return nil, errors.New("unable to find the CoreDNS Corefile data")
+	}
+
+	// Get the current CoreDNS deployment.
+	deployment := &appsv1.Deployment{}
+	if err := w.Client.Get(ctx, corednsKey, deployment); err != nil {
+		return nil, errors.Wrapf(err, "unable to get %s deployment from target cluster", corednsKey.String())
+	}
+
+	// Parse deployment image.
+	addonImage := deployment.Spec.Template.Spec.Containers[0].Image
+	ref, err := reference.ParseNormalizedNamed(addonImage)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse %s deployment image", corednsKey.String())
+	}
+	tagged, hasTag := ref.(reference.Tagged)
+	if !hasTag {
+		return nil, errors.Errorf("%s deployment does not have a valid image tag", corednsKey.String())
+	}
+	parsedTag, err := parseImageTag(tagged.Tag())
+	if err != nil {
+		return nil, err
+	}
+
+	return &coreDNSInfo{
+		Corefile:       corefile,
+		Deployment:     deployment,
+		CurrentVersion: parsedTag,
+	}, nil
+}
+
+// UpdateCoreDNSDeployment will patch the deployment image to the
+// imageRepo:imageTag in the KCP dns. It will also ensure the volume of the
+// deployment uses the Corefile key of the coredns configmap.
+func (w *Workload) updateCoreDNSDeployment(ctx context.Context, info *coreDNSInfo, dns *kubeadmv1.DNS) error {
+	containers := info.Deployment.Spec.Template.Spec.Containers
+	if len(containers) == 0 {
+		return errors.New("failed to update coredns deployment: deployment spec has no containers")
+	}
+
+	// Parse deployment image.
+	image := info.Deployment.Spec.Template.Spec.Containers[0].Image
+	imageRef, err := reference.ParseNormalizedNamed(image)
```

Comment:
Any particular reason to re-parse the image info here, rather than adding the image repository information to the `coreDNSInfo` struct that we are creating in `getCoreDNSInfo`?

Reply:
Refactored this, ptal